File: SemiSupervisedPosTagger/SemiSupervisedPosTagger/averaged_perceptron.cpp
//
// Created by Mohammad Sadegh Rasooli on 1/6/15.
// Copyright (c) 2015 Mohammad Sadegh Rasooli. All rights reserved.
//

#include <iostream>
#include <fstream>
#include "averaged_perceptron.h"
#include "assert.h"

averaged_perceptron::averaged_perceptron(const int tag_size, const int feat_size) {
    weights = new unordered_map<int, float> *[tag_size];
    averaged_weights = new unordered_map<int, float> *[tag_size];
    for (int i = 0; i < tag_size; i++) {
        weights[i] = new unordered_map<int, float>[feat_size];
        averaged_weights[i] = new unordered_map<int, float>[feat_size];
    }
    iteration = 1;
    this->tag_size=tag_size;
    this->feat_size=feat_size;
}

averaged_perceptron::averaged_perceptron(const int tag_size, const int feat_size, unordered_map<int, float> **averaged_weights) {
    weights = new unordered_map<int, float> *[tag_size];
    for (int i = 0; i < tag_size; i++)
        weights[i] = new unordered_map<int, float>[feat_size];
    iteration = 1;
    this->averaged_weights = averaged_weights;
    this->tag_size=tag_size;
    this->feat_size=feat_size;
}

averaged_perceptron *averaged_perceptron::load_model(const char *file_path) {
    ifstream reader(file_path);
    if(!reader)
        cout<<"file does not exist!"<<endl;

    int tag_size, feat_size;
    reader >> tag_size;
    reader >> feat_size;

    unordered_map<int, float>** averaged_weights = new unordered_map<int, float> *[tag_size];
    for (int i = 0; i < tag_size; i++) {
        averaged_weights[i] = new unordered_map<int, float>[feat_size];
    }

    int t_index, f_index;
    int feat;
    float value;
    while (!reader.eof()) {
        reader >> t_index;
        reader  >> f_index;
        reader>> feat;
        reader >> value;
        averaged_weights[t_index][f_index][feat]=value;
    }
    reader.close();
    return new averaged_perceptron(tag_size, feat_size,averaged_weights);
}

void averaged_perceptron::save_model(const char *file_path) {
    ofstream writer;
    writer.open(file_path,ios::binary|ios::out);
    writer<<tag_size<<endl;
    writer<<feat_size<<endl;

    for(int i=0;i<tag_size;i++) {
        for (int j = 0; j < feat_size; j++) {
            for(auto kv:weights[i][j]){
                float new_val=kv.second-(averaged_weights[i][j][kv.first]/iteration);
                if (new_val!=0.0f) {
                    writer << i << endl;
                    writer << j << endl;
                    writer << kv.first << endl;
                    writer << new_val << endl;
                }
            }
        }
    }
    writer.flush();
    writer.close();
}

void averaged_perceptron::increment_iteration() {
    iteration++;
}

void averaged_perceptron::change_weight(const int tag_index, const int feat_index, const int feature, const float change) {
	if(feature==-1)
		return;

    if(weights[tag_index][feat_index].count(feature)>0)
        weights[tag_index][feat_index][feature]=weights[tag_index][feat_index][feature]+change;
    else
        weights[tag_index][feat_index][feature]=change;

    if(averaged_weights[tag_index][feat_index].count(feature)>0)
        averaged_weights[tag_index][feat_index][feature]=averaged_weights[tag_index][feat_index][feature]+iteration*change;
    else
        averaged_weights[tag_index][feat_index][feature]=iteration*change;
}

// returns the score of the features given the tag index; arr_size is for checking the size of the array
float averaged_perceptron::score(vector<int> const features, const int tag_index, const bool is_decode) {
    float score=0;
    unordered_map<int,float>* map=is_decode?averaged_weights[tag_index]:weights[tag_index];

    for(int i=0;i<features.size();i++){
	    int feat=features.at(i);
        if(feat!=-1 && map[i].count(feat)>0)
            score+=map[i][feat];
    }

    return score;
}


unordered_map<int, float> **averaged_perceptron::get_averaged_weights() {
	unordered_map<int,float>** avg  = new unordered_map<int, float> *[tag_size];
	for (int i = 0; i < tag_size; i++) {
		avg[i] = new unordered_map<int, float>[feat_size];
	}

	for(int i=0;i<tag_size;i++){
		for(int j=0;j<feat_size;j++){
			unordered_map<int,float> w=weights[i][j];
			unordered_map<int,float> a=averaged_weights[i][j];

			for(unordered_map<int,float>::iterator iter=w.begin();iter!=w.end();++iter){
				int key=iter->first;
				float val=iter->second;
				float a_val=a[key];

				float new_val=val-(a_val/iteration);
				if(new_val!=0.0f)
					avg[i][j][key]=new_val;
			}
		}
	}
	return avg;
}

float averaged_perceptron::score(const int tag_index, const int feat_index, const int feat, const bool is_decode) {
	if(feat==-1)
		return 0;
	unordered_map<int,float>* map=is_decode?averaged_weights[tag_index]:weights[tag_index];

	if(map[feat_index].count(feat)>0)
		return map[feat_index][feat];

	return 0;
}

int averaged_perceptron::size() {
    int s=0;
    for(int i=0;i< tag_size;i++)
        for(int j=0;j<feat_size;j++)
            s+=averaged_weights[i][j].size();
    return s;
}

File: src/SemiSupervisedPOSTagger/Learning/AveragedPerceptron.java
/**
 * Copyright 2014, Yahoo! Inc.
 * Licensed under the terms of the Apache License 2.0. See LICENSE file at the project root for terms.
 * Modified by Mohammad Sadegh Rasooli
 */

package SemiSupervisedPOSTagger.Learning;

import SemiSupervisedPOSTagger.Structures.InfoStruct;
import SemiSupervisedPOSTagger.Structures.SpecialWords;

import java.util.HashMap;
import java.util.HashSet;

public class AveragedPerceptron {
    /**
     * This class tries to implement averaged Perceptron algorithm
     * Collins, Michael. "Discriminative training methods for hidden Markov models: Theory and experiments with Perceptron algorithms."
     * In Proceedings of the ACL-02 conference on Empirical methods in natural language processing-Volume 10, pp. 1-8.
     * Association for Computational Linguistics, 2002.
     * <p/>
     * The averaging update is also optimized by using the trick introduced in Hal Daume's dissertation.
     * For more information see the second chapter of his thesis:
     * Harold Charles Daume' III. "Practical Structured SemiSupervisedPOSTagger.Learning Techniques for Natural Language Processing", PhD thesis, ISI USC, 2006.
     * http://www.umiacs.umd.edu/~hal/docs/daume06thesis.pdf
     */

    
    //todo make semi-sparse vectors
    
    /**
     * For the weights for all features
     */
    public  HashMap<Integer, Float>[][] featureWeights;
    public int iteration;
    /**
     * This is the main part of the extension to the original perceptron algorithm which the averaging over all the history
     */
    public HashMap<Integer, Float>[][] averagedWeights;
    
    public HashMap<Integer,Float> penalizerWeight;
    public HashMap<Integer,Float> avgPenalizerWeight;
    
    public HashMap<Integer,HashSet<Integer>> tagDictionary;
    
    public AveragedPerceptron(final int tagSize, final int featSize, HashMap<Integer,HashSet<Integer>> tagDictionary) {
        featureWeights = new HashMap[tagSize][featSize];
        for (int i = 0; i < featureWeights.length; i++)
            for (int j = 0; j < featureWeights[i].length; j++)
                featureWeights[i][j] = new HashMap<Integer, Float>();
        iteration = 1;
        this.averagedWeights = new HashMap[tagSize][featSize];
        for (int i = 0; i < averagedWeights.length; i++)
            for (int j = 0; j < averagedWeights[i].length; j++)
                averagedWeights[i][j] = new HashMap<Integer, Float>();
        this.tagDictionary=tagDictionary;
        
        penalizerWeight=new HashMap<Integer, Float>();
        avgPenalizerWeight=new HashMap<Integer, Float>();
    }

    private AveragedPerceptron(int tagSize, int featSize, HashMap<Integer, Float>[][] averagedWeights,HashMap<Integer,HashSet<Integer>> tagDictionary,HashMap<Integer,Float> avgPenalizerWeight) {
        featureWeights = new HashMap[tagSize][featSize];
        for (int i = 0; i < featureWeights.length; i++)
            for (int j = 0; j < featureWeights[i].length; j++)
                featureWeights[i][j] = new HashMap<Integer, Float>();

        iteration = 1;
        this.averagedWeights = averagedWeights;
        this.tagDictionary=tagDictionary;
        this.avgPenalizerWeight=avgPenalizerWeight;
    }

    public AveragedPerceptron(InfoStruct info) {
         this(info.tagSize,info.featSize,info.averagedWeights,info.tagDictionary,info.penalizerWeight);
    }

    public float changeWeight(int tagIndex,int featIndex, int featureName, float change) {
        if (featureName == -1)
            return 0;
       if(featIndex==featureSize()-1){
            if(!penalizerWeight.containsKey(featureName))
                penalizerWeight.put(featureName,0f);
           penalizerWeight.put(featureName,penalizerWeight.get(featureName)+change);
           
           if(!avgPenalizerWeight.containsKey(featureName))
               avgPenalizerWeight.put(featureName,0f);
           avgPenalizerWeight.put(featureName,avgPenalizerWeight.get(featureName)+change);
       }//   else {
           // System.out.println(tagIndex+" "+featIndex+" "+featureName+" -> "+change);
           HashMap<Integer, Float> map = featureWeights[tagIndex][featIndex];
           Float value = map.get(featureName);
           if (value != null)
               map.put(featureName, change + value);
           else
               map.put(featureName, change);

           map = averagedWeights[tagIndex][featIndex];
           value = map.get(featureName);
           if (value != null)
               map.put(featureName, (iteration * change) + value);
           else
               map.put(featureName, iteration * change);
      // }
        return change;
    }

    /**
     * Adds to the iterations
     */
    public void incrementIteration() {
        iteration++;
    }

    /**
     * Returns the score of the specific feature
     *
     * @param features the features in the current instance
     * @return
     */
    public float score(final int[] features, int tagIndex, boolean isDecode) {
        float score = 0;
        final HashMap<Integer, Float>[] weights;
        if (!isDecode) {
            weights = featureWeights[tagIndex];

        } else {
            weights = averagedWeights[tagIndex];
        }

        for (int i = 0; i < features.length; i++) {
       //     if(features[i]== SpecialWords.unknown.value)
         //       continue;
            Float value = (weights[i]).get(features[i]);

            if (value != null)
                score += value;
        }
        return score;
    }

    public float score(final int tagIndex, final int featIndex, final int feat, final boolean isDecode) {
      float score=0f;
        //if (feat == SpecialWords.unknown.value)
       //     return score;
       if(featIndex==featureSize()-1){
            if(isDecode){
                 if(avgPenalizerWeight.containsKey(feat))
                     score=   avgPenalizerWeight.get(feat);
            }   else {
                if(penalizerWeight.containsKey(feat))
                    score=   penalizerWeight.get(feat);
            }
       } // else {

           HashMap<Integer, Float> map = isDecode ? averagedWeights[tagIndex][featIndex] : featureWeights[tagIndex][featIndex];

           if (map.containsKey(feat))
               score+= map.get(feat);
     //  }
        return score;
    }

    public int size() {
        int size = 0;
        for (int i = 0; i < averagedWeights.length; i++)
            for (int j = 0; j < averagedWeights[i].length; j++)
                size += (averagedWeights[i][j]).size();
        return size;
    }

    public int featureSize() {
        return averagedWeights[0].length;
    }

    public int tagSize() {
        return averagedWeights.length;
    }

    public HashMap<Integer, Float>[][] getAveragedWeights() {
        HashMap<Integer, Float>[][] avg=new HashMap[tagSize()][featureSize()];
        for(int i=0;i<tagSize();i++){
            for(int j=0;j<featureSize();j++){
                HashMap<Integer,Float> w=featureWeights[i][j];
                HashMap<Integer,Float> a=averagedWeights[i][j];
                avg[i][j]=new HashMap<Integer, Float>();

                for(int key:w.keySet()){
                    float val=w.get(key);
                    float aVal=a.get(key);
                    float newVal=val-(aVal/iteration);
                    if(newVal!=0.0f)
                        avg[i][j].put(key,newVal);
                }
            }
        }
        return avg;
    }
    
    public int dictCondition(int word, int tag){
        int cond=0;
        if(tagDictionary.containsKey(word)){
            cond=0;
            if(tagDictionary.get(word).contains(tag))
                cond=1;
        }
        return cond;
    }

    public HashMap<Integer, Float> getAvgPenalizerWeight() {
      HashMap<Integer,Float> avgP=new HashMap<Integer, Float>();
        for(int f:penalizerWeight.keySet()){
            float val=penalizerWeight.get(f)-(avgPenalizerWeight.get(f)/iteration);
            if(val!=0f)
                avgP.put(f,val);
        }
      
        return avgP;
    }
}